### PR TITLE
[surface head node bootstrap errors] trim error message before passing to cfn-signal

### PR DESCRIPTION
**Pull request by delongmeng**
### Description of changes
Error messages written into the dedicated error message file during head node bootstrap process will be sent to cfn-signal. However, there is a [size limit of 4096 bytes for cfn-signal](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cloudformation-limits.html). If it exceeds the limit, the CouldFormation stack will show "Unparsable WaitCondition data". Since there is a chance the error message can exceed the size limit, we need to handle this situation. 

### Tests
Manually tested.  
* before code change:
  - Manually store long message in the error message file and CloudFormation stack shows "Unparsable WaitCondition data".
 
* after code change:
  - Manually store long message in the error message file and CloudFormation stack shows trimmed message.
  - Create a cluster with error in custom script (error message will be short enough) and can still receive the same error message from CloudFormation stack.

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/resources/head_node/user_data.sh [hunk 1]
```diff
@@ -66,7 +66,8 @@ function error_exit
 {
   # wait logs flush before signaling the failure
   sleep 10
-  reason=$(cat /var/log/parallelcluster/headnode_bootstrap_error_msg 2>/dev/null) || reason="$1"
+  cutoff=$(expr 4096 - $(stat --printf="%s" /tmp/wait_condition_handle.txt))
```
Comment (chenwany) [inline cli/src/pcluster/resources/head_node/user_data.sh:71]: Could you add a comment about why do we cut off the log?
Is there any restriction for some special characters on reason? What happen if some special charater show up in the failure result? Is the logs able to show up in stack reason?

Comment (delongmeng) [reply cli/src/pcluster/resources/head_node/user_data.sh:71]: good point. added comments to make it clearer.
will need to do some research on special character restriction. Do you happen to know any example and how we handle that in the past? Thanks

Comment (delongmeng) [reply cli/src/pcluster/resources/head_node/user_data.sh:71]: Tested ASCII special characters such as ~  :  '  +  [  \  @  ^  {  %  (  -  \"  *  |  ||  ,  &  &&  <  `  }  .  _  =  ]  !  >  ;  ?  #  $  )  /, and they all show up in the CloudFormation stack without any issues.
Tested non-ASCII unicode characters,  such as (1) – (2) “ (3) ” (4) 中文, and they were all converted to question marks ((1) ? (2) ? (3) ? (4) ??), without affecting other characters (so the overall error message is still very well readable).
Does this address the concern?

Comment (delongmeng) [reply cli/src/pcluster/resources/head_node/user_data.sh:71]: Information in the error message file:

> Failed to run postinstall; Testing special characters ~ : ' + [ \ @ ^ { % ( - " * | || , & && < ` } . _ = ] ! > ; ? # $ ) / ; Testing some non-ascii characters (1) – (2) “ (3) ” (4) 中文; End of Testing.


Information in CloudFormation stack:

> Failed to run postinstall; Testing special characters ~ : ' + [ \ @ ^ { % ( - " * | || , & && < ` } . _ = ] ! > ; ? # $ ) / ; Testing some non-ascii characters (1) ? (2) ? (3) ? (4) ??; End of Testing.

**Review by enrico-usai — APPROVED** [review]

**Pull request merged**